File: ui/src/lib/fiscal/regimenes-fiscales.ts
/**
 * Catálogo de regímenes fiscales SAT (c_RegimenFiscal del Anexo 20 CFDI 4.0).
 * Solo los regímenes vigentes que un contribuyente puede tener configurados.
 */

export interface RegimenFiscalCatalogo {
  clave: string;
  descripcion: string;
  tipoPersona: 'PF' | 'PM' | 'ambos';
}

export const REGIMENES_FISCALES: RegimenFiscalCatalogo[] = [
  // Personas Morales
  { clave: '601', descripcion: 'General de Ley Personas Morales', tipoPersona: 'PM' },
  { clave: '603', descripcion: 'Personas Morales con Fines no Lucrativos', tipoPersona: 'PM' },
  { clave: '620', descripcion: 'Sociedades Cooperativas de Producción', tipoPersona: 'PM' },
  { clave: '622', descripcion: 'Actividades Agrícolas, Ganaderas, Silvícolas y Pesqueras', tipoPersona: 'PM' },
  { clave: '623', descripcion: 'Opcional para Grupos de Sociedades', tipoPersona: 'PM' },
  { clave: '624', descripcion: 'Coordinados', tipoPersona: 'PM' },

  // Personas Físicas
  { clave: '605', descripcion: 'Sueldos y Salarios e Ingresos Asimilados a Salarios', tipoPersona: 'PF' },
  { clave: '606', descripcion: 'Arrendamiento', tipoPersona: 'PF' },
  { clave: '607', descripcion: 'Régimen de Enajenación o Adquisición de Bienes', tipoPersona: 'PF' },
  { clave: '608', descripcion: 'Demás Ingresos', tipoPersona: 'PF' },
  { clave: '610', descripcion: 'Residentes en el Extranjero sin Establecimiento Permanente en México', tipoPersona: 'PF' },
  { clave: '611', descripcion: 'Ingresos por Dividendos (socios y accionistas)', tipoPersona: 'PF' },
  { clave: '612', descripcion: 'Personas Físicas con Actividades Empresariales y Profesionales', tipoPersona: 'PF' },
  { clave: '614', descripcion: 'Ingresos por Intereses', tipoPersona: 'PF' },
  { clave: '615', descripcion: 'Régimen de los Ingresos por Obtención de Premios', tipoPersona: 'PF' },
  { clave: '616', descripcion: 'Sin Obligaciones Fiscales', tipoPersona: 'PF' },
  { clave: '621', descripcion: 'Incorporación Fiscal', tipoPersona: 'PF' },
  { clave: '629', descripcion: 'De los Regímenes Fiscales Preferentes y de las Empresas Multinacionales', tipoPersona: 'PF' },

  // Ambos
  { clave: '625', descripcion: 'Régimen de las Actividades Empresariales con Ingresos a través de Plataformas Tecnológicas', tipoPersona: 'ambos' },
  { clave: '626', descripcion: 'Régimen Simplificado de Confianza', tipoPersona: 'ambos' },
];

export function getRegimenesByTipoPersona(
  tipoPersona: 'PF' | 'PM',
): RegimenFiscalCatalogo[] {
  return REGIMENES_FISCALES.filter(
    (r) => r.tipoPersona === tipoPersona || r.tipoPersona === 'ambos',
  );
}

/** Clave del Régimen Simplificado de Confianza (RESICO), PF y PM. */
export const CLAVE_RESICO = '626';

/**
 * Regímenes que por DEFAULT no presentan DIOT (la obligación del art. 32
 * fracc. VIII LIVA sigue al IVA, no al régimen de ISR). Mapeo investigado
 * contra RMF 2025/2026 y RFA 2025 — detalle y fundamentos en
 * docs/producto/diot-2025.md («¿Quién presenta la DIOT?»):
 *
 * - 626 RESICO (PF y PM): relevados por la regla 3.13.19 RMF.
 * - 605, 607, 608, 610, 611, 614, 615, 616, 629: no son sujetos del IVA por
 *   esos ingresos (sueldos, enajenación accidental, dividendos, intereses,
 *   premios, sin obligaciones, REFIPRES).
 * - 621 RIF: relevado si informa proveedores en la bimestral (art. 5-E LIVA
 *   + art. 23 LIF).
 * - 603, 606, 625: condicionales con default NO (no lucrativas sin actos
 *   gravados; arrendamiento — casa habitación exenta y relevación ≤ $4 MDP
 *   de la regla 2.8.1.17 RMF; plataformas con retención definitiva del
 *   art. 18-M LIVA o ≤ $4 MDP).
 *
 * Los que default SÍ: 601, 612 (obligado salvo ingresos ≤ $4 MDP — el
 * usuario lo apaga con el toggle), 620, 622 (facilidad semestral), 623, 624
 * (el coordinado puede presentarla global). El toggle `presenta_diot` de la
 * configuración de la empresa cubre todos los casos condicionales.
 */
const REGIMENES_SIN_DIOT_DEFAULT = new Set([
  '603', '605', '606', '607', '608', '610', '611', '614', '615', '616',
  '621', '625', '626', '629',
]);

/** ¿Este régimen (clave individual) obliga a la DIOT por regla general? */
export function regimenPresentaDiot(clave: string): boolean {
  return !REGIMENES_SIN_DIOT_DEFAULT.has(clave);
}

/**
 * Derivación por régimen: presenta DIOT si ALGÚN régimen configurado la trae
 * por default. Sin regímenes configurados (o clave desconocida) se asume que
 * sí presenta — es el default seguro; omitirla estando obligado genera multas.
 */
export function regimenesPresentanDiot(
  regimenes?: { clave: string }[] | null,
): boolean {
  if (!regimenes || regimenes.length === 0) return true;
  return regimenes.some((r) => regimenPresentaDiot(r.clave));
}

/**
 * ¿La empresa presenta DIOT? El override manual (`presenta_diot`, toggle en la
 * configuración de la empresa) manda; sin override se deriva del régimen.
 */
export function empresaPresentaDiot(
  empresa?: {
    regimenes_fiscales?: { clave: string }[];
    presenta_diot?: boolean | null;
  } | null,
): boolean {
  if (typeof empresa?.presenta_diot === 'boolean') return empresa.presenta_diot;
  return regimenesPresentanDiot(empresa?.regimenes_fiscales);
}

File: sat_descarga/utils/regimenes_fiscales.py
"""
Catálogo de regímenes fiscales SAT (c_RegimenFiscal del Anexo 20 CFDI 4.0).

Archivo gemelo de `ui/src/lib/fiscal/regimenes-fiscales.ts` — mantener ambos en
espejo al agregar/retirar regímenes. Este lado lo consume el parser de la
Constancia de Situación Fiscal para mapear la descripción impresa en el PDF a
su clave de catálogo.
"""

from __future__ import annotations

import re
import unicodedata
from typing import Optional

# {clave, descripcion, tipo_persona: "PF" | "PM" | "ambos"}
REGIMENES_FISCALES: list[dict] = [
    # Personas Morales
    {"clave": "601", "descripcion": "General de Ley Personas Morales", "tipo_persona": "PM"},
    {"clave": "603", "descripcion": "Personas Morales con Fines no Lucrativos", "tipo_persona": "PM"},
    {"clave": "620", "descripcion": "Sociedades Cooperativas de Producción", "tipo_persona": "PM"},
    {"clave": "622", "descripcion": "Actividades Agrícolas, Ganaderas, Silvícolas y Pesqueras", "tipo_persona": "PM"},
    {"clave": "623", "descripcion": "Opcional para Grupos de Sociedades", "tipo_persona": "PM"},
    {"clave": "624", "descripcion": "Coordinados", "tipo_persona": "PM"},

    # Personas Físicas
    {"clave": "605", "descripcion": "Sueldos y Salarios e Ingresos Asimilados a Salarios", "tipo_persona": "PF"},
    {"clave": "606", "descripcion": "Arrendamiento", "tipo_persona": "PF"},
    {"clave": "607", "descripcion": "Régimen de Enajenación o Adquisición de Bienes", "tipo_persona": "PF"},
    {"clave": "608", "descripcion": "Demás Ingresos", "tipo_persona": "PF"},
    {"clave": "610", "descripcion": "Residentes en el Extranjero sin Establecimiento Permanente en México", "tipo_persona": "PF"},
    {"clave": "611", "descripcion": "Ingresos por Dividendos (socios y accionistas)", "tipo_persona": "PF"},
    {"clave": "612", "descripcion": "Personas Físicas con Actividades Empresariales y Profesionales", "tipo_persona": "PF"},
    {"clave": "614", "descripcion": "Ingresos por Intereses", "tipo_persona": "PF"},
    {"clave": "615", "descripcion": "Régimen de los Ingresos por Obtención de Premios", "tipo_persona": "PF"},
    {"clave": "616", "descripcion": "Sin Obligaciones Fiscales", "tipo_persona": "PF"},
    {"clave": "621", "descripcion": "Incorporación Fiscal", "tipo_persona": "PF"},
    {"clave": "629", "descripcion": "De los Regímenes Fiscales Preferentes y de las Empresas Multinacionales", "tipo_persona": "PF"},

    # Ambos
    {"clave": "625", "descripcion": "Régimen de las Actividades Empresariales con Ingresos a través de Plataformas Tecnológicas", "tipo_persona": "ambos"},
    {"clave": "626", "descripcion": "Régimen Simplificado de Confianza", "tipo_persona": "ambos"},
]

# La CSF antepone "Régimen de/del/de los…" a nombres que el catálogo lista sin
# prefijo (y viceversa: 607/615/625/626 ya lo traen en el catálogo). Se quita
# en ambos lados antes de comparar.
_PREFIJO_REGIMEN = re.compile(r"^regimen\s+(de\s+(los|las|la|el)\s+|de\s+|del\s+)?")


def _normalizar(texto: str) -> str:
    """lowercase, sin acentos, espacios colapsados y sin prefijo 'Régimen de…'."""
    plano = unicodedata.normalize("NFD", texto.lower())
    plano = "".join(c for c in plano if not unicodedata.combining(c))
    plano = " ".join(plano.split())
    return _PREFIJO_REGIMEN.sub("", plano)


_CATALOGO_NORMALIZADO = [(_normalizar(r["descripcion"]), r) for r in REGIMENES_FISCALES]


def clave_de_regimen(descripcion_csf: str) -> Optional[str]:
    """
    Mapea la descripción de un régimen tal como viene en la CSF a su clave del
    catálogo. Tolerante a acentos, mayúsculas y al prefijo "Régimen de…".
    Devuelve None si no hay match razonable.
    """
    objetivo = _normalizar(descripcion_csf)
    if not objetivo:
        return None

    for normalizada, regimen in _CATALOGO_NORMALIZADO:
        if normalizada == objetivo:
            return regimen["clave"]

    # Containment bidireccional; si varios matchean gana la descripción más
    # larga del catálogo (la más específica).
    candidatos = [
        (len(normalizada), regimen)
        for normalizada, regimen in _CATALOGO_NORMALIZADO
        if normalizada in objetivo or objetivo in normalizada
    ]
    if candidatos:
        return max(candidatos, key=lambda c: c[0])[1]["clave"]
    return None
